impeller/entity/contents/filters/runtime_effect_filter_contents.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/entity/contents/filters/runtime_effect_filter_contents.h"
+
+#include <cstring>
+#include <optional>
+
+#include "impeller/base/validation.h"
+#include "impeller/entity/contents/anonymous_contents.h"
+#include "impeller/entity/contents/runtime_effect_contents.h"
+#include "impeller/geometry/size.h"
+
+namespace impeller {
+
+void RuntimeEffectFilterContents::SetRuntimeStage(
+    std::shared_ptr<RuntimeStage> runtime_stage) {
+  runtime_stage_ = std::move(runtime_stage);
+}
+
+void RuntimeEffectFilterContents::SetUniforms(
+    std::shared_ptr<std::vector<uint8_t>> uniforms) {
+  uniforms_ = std::move(uniforms);
+}
+
+void RuntimeEffectFilterContents::SetTextureInputs(
+    std::vector<RuntimeEffectContents::TextureInput> texture_inputs) {
+  texture_inputs_ = std::move(texture_inputs);
+}
+
+// |FilterContents|
+std::optional<Entity> RuntimeEffectFilterContents::RenderFilter(
+    const FilterInput::Vector& inputs,
+    const ContentContext& renderer,
+    const Entity& entity,
+    const Matrix& effect_transform,
+    const Rect& coverage,
+    const std::optional<Rect>& coverage_hint) const {
+  if (inputs.empty()) {
+    return std::nullopt;
+  }
+
+  auto input_snapshot =
+      inputs[0]->GetSnapshot("RuntimeEffectContents", renderer, entity);
+  if (!input_snapshot.has_value()) {
+    return std::nullopt;
+  }
+  std::optional<Rect> maybe_input_coverage = input_snapshot->GetCoverage();
+  if (!maybe_input_coverage.has_value()) {
+    return std::nullopt;
+  }
+  Rect input_coverage = maybe_input_coverage.value();
+  // The shader is required to have at least one sampler, the first of
+  // which is treated as the input and a vec2 size uniform to compute the
+  // offsets. These are validated at the dart:ui layer, but to avoid crashes we
+  // check here too.
+  if (texture_inputs_.size() < 1 || uniforms_->size() < 8) {
+    VALIDATION_LOG
+        << "Invalid fragment shader in RuntimeEffectFilterContents. "
+        << "Shader must have at least one sampler and a vec2 size uniform.";
```

Comment:
How does this manifest for users when they have an invalid shader?  Are they getting dart exceptions, crashes, is this message visible to them?  Is rendering after this aborted?

I think ideally this would not abort any rendering, the error would be visible to the filter being rendered in the UI, the user could hot reload to address the issue.

Rendering a magenta rect with a loud console message would be fine too.

Reply:
I believe we should always detect this in dart:ui and throw an exception when the filter is constructed:

https://github.com/flutter/engine/pull/53490/files#diff-1deac5f38d2c30929d40af54c24c990c604c6b7d7bb21a103d28ffd50c8dcb91R4102

However, we also need to check here since there may eventually be non dart users of this funciton. returning nullopt here causes the filter to not show up but anything that was not rendered as a child of the filter will still show uo.

Reply:
Okay, sounds good to me.  Can we expand the dart test to test these 2 cases individually, uniform size being wrong and texture inputs being wrong?  Right now it just tests them together.

Reply:
Done